text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,234 @@
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The `#[diagnostic]` attribute namespace
+
+This RFC proposes to introduce a new `#[diagnostic]` attribute namespace. This namespace is supposed to contain different attributes, which allow users to hint the compiler to emit specific error messages in certain cases like type mismatches, unsatisfied trait bounds or similar situations. By collecting such attributes in a common namespace it is easier for users to find useful attributes and it is easier for the language team to establish a set of common rules for these attributes.
+
+Attributes in this namespace are generally expected to be formed like:
+```rust
+#[diagnostic::attribute(option)]
+```
+where several `option` entries can appear in the same attribute. `option` is expected to be a valid attribute argument in this position. Attributes in this namespace are versioned. Any incompatible change to an existing stable attribute in this namespace requires bumping the version of the diagnostic attribute namespace. Crates using these attributes must specify the expected version of the diagnostic namespace via a `#![diagnostic::v{version_number}]` crate top-level attribute. If multiple such attributes are present the compiler will choose the highest version number and ignore the rest. The compiler is encouraged to provide a lint for outdated versions of the diagnostic attribute namespace. This mechanism exists so that a third party crate can provide a shim implementation of the corresponding attributes to translate between different versions of the diagnostic attribute namespace on different rust versions.
+
+Any attribute in this namespace may:
+
+* Hint the compiler to emit a specific error message in a specific situation
+* Only affect the messages emitted by a compiler in case of a failed compilation
+
+Any attribute in this namespace is not allowed to:
+
+* Change the result of the compilation, which means applying such an attribute should never cause an error as long as they are syntactically valid
+* pass-through information from the source of the diagnostic in a way that users can rely on. E.g. Such an attribute should not allow users to keep the compilation successful and dump information about `extern` blocks to generate C header files
+
+The compiler is allowed to:
+
+* Ignore the hints provided by:
+    + A specific attribute
+    + A specific option 
+* Change the set of supported attributes and options at any time
+* Provide warning lints for malformed or otherwise not accepted attributes 
+
+The compiler must not:
+
+* Change the semantic of an attribute or option, these kind of change requires a bump of the diagnostic attribute space version number
+* Emit an hard error on malformed attribute, although emitting lints is fine
+
+Adding a new attribute or option to the `#[diagnostic]` namespace is a decision of the compiler team. The envisioned procedure for this is as following:
```

Comment:
> Put another way, is the justification for bypassing T-lang approval essentially that: it doesn't affect semantics of programs, including whether they compile, and that alternate compilers can ignore it?

rustc can ignore them, they are effectively a new kind of comments from the lang perspective, even if we as compiler developers will want to give users a reliable experience.

Reply:
The original proposal (before opening this RFC) was to introduce a `#[rustc::]` attribute namespace. That would make it much clearer that this these attributes are compiler specific. Some members of the compiler team and the language team raised the valid concern that these specific attributes around diagnostics might be relevant for other compilers as well. So a different named attribute namespace seems to be more fitting.

Now this RFC tries to specify the namespace in such a way that supporting anything in that namespace should be optional for other compilers. So if these attributes do not fit their model, they are free to ignore these attributes (or ignore certain options only). Otherwise the main argument for "bypassing" approval from the language team is to prevent spamming the team with otherwise trivial changes. Again that was something that was proposed by @joshtriplett .